inloop/statistics/templates/statistics/solutions_histogram.html

```diff
@@ -0,0 +1,161 @@
+      <option value="168">Last 7 days</option>
+      <option selected value="720">Last 30 days</option>
+      <option value="2160">Last 90 days</option>
+      <option value="8760">Last 365 days</option>
+      <option value="undefined">All time</option>
+    </select>
+  </div>
+
+  <canvas class="mt-3" id="chart" width="400" height="200"></canvas>
+</div>
+
+<script id="execute-on-load">
+(function onLoad() {
+
+  let chart;
+
+  function showHistogram(selectedGranularity, histogram, fromDate, toDate) {
+    let data = {};
+
+    if (histogram.length > 0) {
+      const dateMappedHistogram = histogram.map((e) => ({
+        date: new Date(e.date),
+        count: e.count
+      }));
+      const sortedHistogram = dateMappedHistogram
+        .sort((a, b) => {
+          if (a > b) return 1;
+          if (a < b) return -1;
+          return 0;
+        });
+
+      if (sortedHistogram[0].date > fromDate) {
+        sortedHistogram.unshift({
+          date: fromDate,
+          count: 0
+        })
+      }
+
+      if (sortedHistogram[sortedHistogram.length - 1].date < toDate) {
+        sortedHistogram.push({
+          date: toDate,
+          count: 0
+        })
+      }
+
+      data = sortedHistogram.map((e) => ({t: e.date, y: e.count}));
+    }
+
+    const config = {
+      data: {
+        datasets: [{
+          label: "Submitted solutions",
+          data: data,
+          borderColor: "#007bff",
+          backgroundColor: "rgba(56, 103, 214, 0.5)",
+        }],
+        labels: []
+      },
+      type: "bar",
+      options: {
+        scales: {
+          xAxes: [{
+            type: "time",
+            time: {
+              unit: selectedGranularity
+            },
+            distribution: "linear"
+          }],
+          yAxes: [{
+            gridLines: {
+              drawBorder: false
+            }
+          }]
+        }
+      }
+    };
+
+    if (chart !== undefined) {
+      chart.destroy();
+    }
+    chart = new Chart("chart", config);
+  }
+
+
+  function reloadSolutionsHistogram() {
+    const granularitySelect = document.getElementById("granularity-select");
+    const selectedGranularity = granularitySelect
+      .options[granularitySelect.selectedIndex]
+      .value;
+
+    const dateSpanSelect = document.getElementById("date-span-select");
+    const selectedHoursValue = dateSpanSelect
+      .options[dateSpanSelect.selectedIndex]
+      .value;
+    let selectedHours;
+    if (selectedHoursValue === "undefined") {
+      selectedHours = undefined;
+    } else {
+      selectedHours = parseInt(selectedHoursValue);
+    }
+
+    let fromTimestamp, toTimestamp, fromDate, toDate;
+
+    if (selectedHours !== undefined) {
+      fromDate = new Date()
+        .startOfDay()
+        .minusHours(selectedHours);
+      fromTimestamp = fromDate.toISOString();
+      toDate = new Date()
+        .endOfDay();
+      toTimestamp = toDate.toISOString();
+    }
+
+    const request = new XMLHttpRequest();
```

Comment:
Above, you used the new `fetch(…)` API, why do you use XHR here? Can't this be migrated to `fetch(…)` as well?

Reply:
This came to my mind today as well, but I didn‘t know out of the box how to include the csrf token header in the fetch call. In the course of cleaning up, I will change this to fetch, but have to look up how to do this. (Low effort needed)

Reply:
You actually have two options: a) either include the token as a header (`header` init option): `fetch(url, {headers: {…}})` or b) set the `credentials` init option, to include cookies in the request (which should include the csrf cookie). See <https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch>.

Reply:
I resolved this this morning by using the associated `Request` API.